File: src/observability/logger.ts
export type Logger = {
    debug: (message: string, fields?: LogFields) => void
    info: (message: string, fields?: LogFields) => void
    warn: (message: string, fields?: LogFields) => void
    error: (message: string, fields?: LogFields) => void
}

type LogLevel = 'debug' | 'info' | 'warn' | 'error'
type LogFields = Record<string, unknown>

const levelWeight: Record<LogLevel, number> = {
    debug: 10,
    info: 20,
    warn: 30,
    error: 40,
}

export function createLogger(options: { level: LogLevel }): Logger {
    const shouldWrite = (level: LogLevel) => levelWeight[level] >= levelWeight[options.level]

    const write = (level: LogLevel, message: string, fields: LogFields = {}) => {
        if (!shouldWrite(level)) return
        const event = {
            time: new Date().toISOString(),
            level,
            message,
            ...fields,
        }
        const line = JSON.stringify(event)
        if (level === 'error') {
            console.error(line)
            return
        }
        console.log(line)
    }

    return {
        debug: (message, fields) => write('debug', message, fields),
        info: (message, fields) => write('info', message, fields),
        warn: (message, fields) => write('warn', message, fields),
        error: (message, fields) => write('error', message, fields),
    }
}

File: internal/logging/logger.go
package logging

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
)

type Logger struct {
	level int
}

const (
	levelDebug = 10
	levelInfo  = 20
	levelWarn  = 30
	levelError = 40
)

func New(level string) Logger {
	return Logger{level: parseLevel(level)}
}

func (l Logger) Debug(message string, fields map[string]any) {
	l.write(levelDebug, "debug", message, fields)
}

func (l Logger) Info(message string, fields map[string]any) {
	l.write(levelInfo, "info", message, fields)
}

func (l Logger) Warn(message string, fields map[string]any) {
	l.write(levelWarn, "warn", message, fields)
}

func (l Logger) Error(message string, fields map[string]any) {
	l.write(levelError, "error", message, fields)
}

func (l Logger) write(weight int, level string, message string, fields map[string]any) {
	if weight < l.level {
		return
	}
	event := map[string]any{
		"time":    time.Now().UTC().Format(time.RFC3339Nano),
		"level":   level,
		"message": message,
	}
	for key, value := range fields {
		event[key] = value
	}
	line, _ := json.Marshal(event)
	if level == "error" {
		fmt.Fprintln(os.Stderr, string(line))
		return
	}
	fmt.Println(string(line))
}

func parseLevel(level string) int {
	switch strings.ToLower(strings.TrimSpace(level)) {
	case "debug":
		return levelDebug
	case "warn":
		return levelWarn
	case "error":
		return levelError
	default:
		return levelInfo
	}
}
